external/fv3fit/fv3fit/emulation/thermobasis/thermo.py

```diff
@@ -0,0 +1,105 @@
+from typing import Optional, Sequence
+import dataclasses
+import tensorflow as tf
+from fv3fit.emulation.thermo import (
+    density,
+    relative_humidity,
+    specific_humidity_from_rh,
+    pressure_thickness,
+)
+
+
+class ThermoBasis:
+    """A thermodynamic basis with specific humidity as the prognostic variable"""
```

Comment:
Cool design, I really like having the same interfaces with properties handling conversion based on what class you're using. I have a doc suggestion and a structural suggestion.

I'd suggest updating the description of the base class to describe that it's a container of thermodynamic variables. The base class doesn't do any conversions or indicate a prognostic variable, it just says "I'm going to provide these tensor attributes".

`qc`, `u`, and `v` are a little out of place with the rest of the variables. `qc` is optional, and these three are not used in any of the thermodynamic calculations, raising the question of why they get their attributes while the scalars get a Sequence. As it's written, the ThermoBasis doesn't have a clear way to extend it with new 3D variables or to use it in a model without `u` and `v`. As a result this feels like a container that was written with a specific model in mind, instead of a general purpose class for doing conversions.

If you intend for this to be used in other models, I'd suggest separating the thermodynamic conversion logic from your model-specific choice of variables. For example, you could remove `u`, `v`, `qc`, and `scalars` from ThermoBasis and define a model-specific EmulationData class using an internal ThermoBasis with property methods for the ThermoBasis variables. Or you could use a ThermoBasis container along with another container dataclass for unrelated variables. Or you could add some kind of registration system for non-thermodynamic variables, but this would be less type-safe.

I think the core of this code is pretty applicable to other models, so it would be good to see some kind of refactor making it clear how this can be extended or used in models with no u/v/qc.

Reply:
Thanks for the thoughts. You are right that these classes are hardcoded to the emulation problem, and yes...adding qc and other variables is challenging. In return, this design makes it much easier to inject physical priors that are specific to the emulation use-case. For other problems, it's probably best to just use the functions in `fv3fit.emulation.thermo` directly.

> qc, u, and v are a little out of place with the rest of the variables.

This class contains qc, u, and v because these are all prognostic variables so need some kind of special treatment. For example, qc does not use a residual architecture.

> qc is optional

I did this to support training models that do not predict qc (which is the use-case I started development with).

Reply:
Ultimately, I do think we could make the data container be problem specific/hard-coded, but use Protocol in the losses to allow for static "duck-typing".

Reply:
OK, what I'm hearing is that this class is specific to your model (which is a particular choice of emulation problem). That's fine - my overall comment here is that it's hard to tell what's specific to your model and what's meant to be re-used by others who might be writing different emulators (e.g. an emulator of a trigger function where u/v aren't used) or non-emulator keras models (is there much of a technical fv3fit-ML distinction between emulator and non-emulator?).

This isn't purely a comment about this PR, the thermodynamic functions you referred to would also be pretty useful in any keras model, not just emulation. But it's hard to discover those re-usable parts because they're in the same namespaces as more private model-specific code (especially when that model-specific code has general-sounding names).

I get a little confused when we talk about "the emulation problem" requiring these prognostic variables, because there is not only a single emulation problem. Are you talking about FullPhysics emulation?

Reply:
Are these just general comments or is there some specific request? 

Andre and I created the `fv3fit.emulation` subpackage because we couldn't think of another name, but wanted to de-couple our work from the existing `fv3fit` code. After all, this code was written without using fv3fit at all. We also considered making a completely new package, but thought it best to concentrate all tensorflow related code in fv3fit. Integrating this package better will require some more thought---e.g. the flat namespace in `fv3fit` can be laborious e.g. sometimes it's nice to use `fv3fit.emulation.online.Config` instead of ` fv3fit.OnlineEmulationConfig`. Documentation/asking a teammate seems like our standard path to improving discoverability.


> Are you talking about FullPhysics emulation?

Yes, but this could be used for non-full physics emulation too, by writing a model which simply copies the copies the input u/v to the output.

Reply:
> Are these just general comments or is there some specific request?

> If you have any ideas on how to improve this I'd appreciate it, though there's no specific PR-blocking change I'm asking for here.

OK, it sounds like we have three development concepts at the same level then - sklearn, keras, and emulation. emulation is like an alternative of keras which was done for emulation. What I'd like to avoid is having `keras` and `emulation` depend on each other at this level, without completely integrating them. I think moving forward if we use code from `emulation` in a model under `keras`, it would make sense to first move that code to `_shared`.

Everything looks good on this for this PR.

Reply:
> I think moving forward if we use code from emulation in a model under keras, it would make sense to first move that code to _shared.

Agreed.